File: skalpai/sdk-node/src/http.ts
import { metrics } from '@opentelemetry/api';
import type { Attributes } from '@opentelemetry/api';
import type { IncomingMessage, ServerResponse } from 'node:http';

export interface HTTPMiddlewareConfig {
  serviceName?: string;
  routeExtractor?: (req: IncomingMessage) => string | undefined;
}

export type NodeHTTPRequestHandler<TReq extends IncomingMessage = IncomingMessage, TRes extends ServerResponse = ServerResponse> =
  (req: TReq, res: TRes) => void | Promise<void>;

type HTTPInstruments = {
  requestCount: ReturnType<ReturnType<typeof metrics.getMeter>['createCounter']>;
  requestDuration: ReturnType<ReturnType<typeof metrics.getMeter>['createHistogram']>;
  activeRequests: ReturnType<ReturnType<typeof metrics.getMeter>['createUpDownCounter']>;
  responseSize: ReturnType<ReturnType<typeof metrics.getMeter>['createHistogram']>;
};

const instrumentsByService = new Map<string, HTTPInstruments>();

export function wrapHttpHandler<TReq extends IncomingMessage, TRes extends ServerResponse>(
  handler: NodeHTTPRequestHandler<TReq, TRes>,
  config: HTTPMiddlewareConfig = {},
): NodeHTTPRequestHandler<TReq, TRes> {
  const serviceName = resolveServiceName(config.serviceName);
  const instruments = getInstruments(serviceName);

  return (req, res) => {
    const start = process.hrtime.bigint();
    const baseAttributes = requestAttributes(req, config.routeExtractor);
    let bytesWritten = 0;
    let settled = false;

    instruments.activeRequests.add(1, baseAttributes);

    const originalWrite = res.write.bind(res);
    const originalEnd = res.end.bind(res);

    res.write = ((chunk: unknown, ...args: unknown[]) => {
      bytesWritten += byteLength(chunk, toEncoding(args[0]));
      return originalWrite(chunk as never, ...(args as never[]));
    }) as typeof res.write;

    res.end = ((chunk?: unknown, ...args: unknown[]) => {
      if (chunk !== undefined) {
        bytesWritten += byteLength(chunk, toEncoding(args[0]));
      }
      return originalEnd(chunk as never, ...(args as never[]));
    }) as typeof res.end;

    const finalize = () => {
      if (settled) return;
      settled = true;

      const durationSeconds = Number(process.hrtime.bigint() - start) / 1e9;
      const finalAttributes = requestAttributes(req, config.routeExtractor, res.statusCode);

      instruments.activeRequests.add(-1, baseAttributes);
      instruments.requestCount.add(1, finalAttributes);
      instruments.requestDuration.record(durationSeconds, finalAttributes);
      instruments.responseSize.record(bytesWritten, finalAttributes);
    };

    res.once('finish', finalize);
    res.once('close', finalize);

    return handler(req, res);
  };
}

function getInstruments(serviceName: string): HTTPInstruments {
  const existing = instrumentsByService.get(serviceName);
  if (existing) return existing;

  const meter = metrics.getMeter(serviceName);
  const instruments: HTTPInstruments = {
    requestCount: meter.createCounter('http.server.request.count', {
      description: 'Total number of HTTP server requests.',
      unit: '{request}',
    }),
    requestDuration: meter.createHistogram('http.server.request.duration', {
      description: 'Total HTTP server request duration in seconds.',
      unit: 's',
    }),
    activeRequests: meter.createUpDownCounter('http.server.active_requests', {
      description: 'Current number of in-flight HTTP server requests.',
      unit: '{request}',
    }),
    responseSize: meter.createHistogram('http.server.response.size', {
      description: 'HTTP response size in bytes.',
      unit: 'By',
    }),
  };
  instrumentsByService.set(serviceName, instruments);
  return instruments;
}

function requestAttributes(
  req: IncomingMessage,
  routeExtractor?: (req: IncomingMessage) => string | undefined,
  statusCode?: number,
): Attributes {
  const attributes: Attributes = {
    'http.request.method': req.method ?? 'GET',
    'server.address': req.headers.host ?? '',
  };

  const route = routeExtractor?.(req);
  if (route) {
    attributes['http.route'] = route;
  }
  if (statusCode && statusCode > 0) {
    attributes['http.response.status_code'] = statusCode;
    attributes['http.response.status_class'] = `${Math.floor(statusCode / 100)}xx`;
  }

  return attributes;
}

function resolveServiceName(serviceName?: string) {
  return serviceName
    || process.env.SKALPAI_SERVICE
    || process.env.OTEL_SERVICE_NAME
    || process.env.npm_package_name
    || 'unknown';
}

function byteLength(chunk: unknown, encoding?: BufferEncoding) {
  if (typeof chunk === 'string') {
    return Buffer.byteLength(chunk, encoding);
  }
  if (Buffer.isBuffer(chunk)) {
    return chunk.length;
  }
  if (chunk instanceof Uint8Array) {
    return chunk.byteLength;
  }
  return 0;
}

function toEncoding(value: unknown): BufferEncoding | undefined {
  return typeof value === 'string' ? value as BufferEncoding : undefined;
}

File: skalpai/sdk-go/http.go
package skalpai

import (
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	metricapi "go.opentelemetry.io/otel/metric"
)

// HTTPMiddlewareConfig controls the HTTP server telemetry emitted by the SDK.
type HTTPMiddlewareConfig struct {
	ServiceName    string
	RouteExtractor func(*http.Request) string
}

type httpServerInstruments struct {
	requestCount    metricapi.Int64Counter
	requestDuration metricapi.Float64Histogram
	activeRequests  metricapi.Int64UpDownCounter
	responseSize    metricapi.Int64Histogram
}

var (
	httpServerMu                   sync.Mutex
	httpServerInstrumentsByService = map[string]httpServerInstruments{}
)

// NewHTTPMiddleware returns a standard net/http middleware that records
// normalized request count, duration, active request, and response size metrics.
func NewHTTPMiddleware(cfg HTTPMiddlewareConfig) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return WrapHTTPHandler(next, cfg)
	}
}

// WrapHTTPHandler wraps a net/http handler with Skalpai HTTP server metrics.
func WrapHTTPHandler(next http.Handler, cfg HTTPMiddlewareConfig) http.Handler {
	if next == nil {
		next = http.NotFoundHandler()
	}

	serviceName := resolveServiceName(cfg.ServiceName)
	instruments := getHTTPServerInstruments(serviceName)

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		baseAttrs := requestAttributes(r, cfg.RouteExtractor, 0)
		instruments.activeRequests.Add(r.Context(), 1, metricapi.WithAttributes(baseAttrs...))
		defer instruments.activeRequests.Add(r.Context(), -1, metricapi.WithAttributes(baseAttrs...))

		startedAt := time.Now()
		recorder := &statusRecorder{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(recorder, r)

		finalAttrs := requestAttributes(r, cfg.RouteExtractor, recorder.status)
		instruments.requestCount.Add(r.Context(), 1, metricapi.WithAttributes(finalAttrs...))
		instruments.requestDuration.Record(r.Context(), time.Since(startedAt).Seconds(), metricapi.WithAttributes(finalAttrs...))
		instruments.responseSize.Record(r.Context(), recorder.bytes, metricapi.WithAttributes(finalAttrs...))
	})
}

func getHTTPServerInstruments(serviceName string) httpServerInstruments {
	httpServerMu.Lock()
	defer httpServerMu.Unlock()

	if instruments, ok := httpServerInstrumentsByService[serviceName]; ok {
		return instruments
	}

	meter := otel.Meter(serviceName)
	requestCount, _ := meter.Int64Counter(
		"http.server.request.count",
		metricapi.WithDescription("Total number of HTTP server requests."),
		metricapi.WithUnit("{request}"),
	)
	requestDuration, _ := meter.Float64Histogram(
		"http.server.request.duration",
		metricapi.WithDescription("Total HTTP server request duration in seconds."),
		metricapi.WithUnit("s"),
	)
	activeRequests, _ := meter.Int64UpDownCounter(
		"http.server.active_requests",
		metricapi.WithDescription("Current number of in-flight HTTP server requests."),
		metricapi.WithUnit("{request}"),
	)
	responseSize, _ := meter.Int64Histogram(
		"http.server.response.size",
		metricapi.WithDescription("HTTP response size in bytes."),
		metricapi.WithUnit("By"),
	)

	instruments := httpServerInstruments{
		requestCount:    requestCount,
		requestDuration: requestDuration,
		activeRequests:  activeRequests,
		responseSize:    responseSize,
	}
	httpServerInstrumentsByService[serviceName] = instruments
	return instruments
}

func resolveServiceName(serviceName string) string {
	if trimmed := strings.TrimSpace(serviceName); trimmed != "" {
		return trimmed
	}
	if env := strings.TrimSpace(os.Getenv("SKALPAI_SERVICE")); env != "" {
		return env
	}
	if env := strings.TrimSpace(os.Getenv("OTEL_SERVICE_NAME")); env != "" {
		return env
	}
	return "unknown"
}

func requestAttributes(r *http.Request, routeExtractor func(*http.Request) string, statusCode int) []attribute.KeyValue {
	attrs := []attribute.KeyValue{
		attribute.String("http.request.method", r.Method),
		attribute.String("server.address", r.Host),
	}
	if routeExtractor != nil {
		if route := strings.TrimSpace(routeExtractor(r)); route != "" {
			attrs = append(attrs, attribute.String("http.route", route))
		}
	}
	if statusCode > 0 {
		attrs = append(attrs,
			attribute.Int("http.response.status_code", statusCode),
			attribute.String("http.response.status_class", statusClass(statusCode)),
		)
	}
	return attrs
}

func statusClass(statusCode int) string {
	if statusCode <= 0 {
		return "unknown"
	}
	return strconv.Itoa(statusCode/100) + "xx"
}

type statusRecorder struct {
	http.ResponseWriter
	status int
	bytes  int64
}

func (r *statusRecorder) WriteHeader(statusCode int) {
	r.status = statusCode
	r.ResponseWriter.WriteHeader(statusCode)
}

func (r *statusRecorder) Write(p []byte) (int, error) {
	if r.status == 0 {
		r.status = http.StatusOK
	}
	n, err := r.ResponseWriter.Write(p)
	r.bytes += int64(n)
	return n, err
}
